Clear notifications after opening a report

arosiclair opened the pull request:
### Details

- Clears notifications for a report when you open it
- Clean up for old unused code in BrowserNotification & LocalNotification libs
- CAVEAT: On web, refreshing the page clears the in-memory notification cache which means this functionality does not work after a page refresh

### Fixed Issues
$ Part of https://github.com/Expensify/Expensify/issues/258535

### Tests

_No tests on mWeb_

1. Log in
4. Accept prompt for notification permission
2. Open concierge chat
3. Send 3 messages to yourself
5. Verify notifications display for the messages
6. Switch to the chat
7. Verify those notifications are cleared
8. Switch away from the app
    - On web/desktop, switch tabs or minimize the window
    - On native, go to the home screen
9. Send another 3 messages to yourself
10. Switch back to the app
11. Verify those notifications are cleared
12. Switch to the concierge chat
13. Send another 3 messages to yourself
14. Switch to _other_ chats (not the one with the messages)
15. Verify those notifications are NOT cleared 

### Offline tests
None

### QA Steps
Same as Tests

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

![image](https://github.com/Expensify/App/assets/6833644/9ca2dbea-7afb-4003-8282-b9cb211b4ca0)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->
![Screenshot_20231128-161044](https://github.com/Expensify/App/assets/6833644/dd4550ce-4386-4137-a79b-548a42fa6651)

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/6833644/80ab9889-d709-4cae-9bcd-cdc92849fea7

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->
![IMG_32274EBE93FC-1](https://github.com/Expensify/App/assets/6833644/45eb872f-4009-4520-86d0-e4f4722bfd58)

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

![Screenshot 2023-12-07 at 12 06 02 PM](https://github.com/Expensify/App/assets/6833644/31e46bc4-a642-46d6-b06f-d5a61f92d046)


</details>

<details>
<summary>MacOS: Desktop</summary>

![Screenshot 2023-12-07 at 12 22 35 PM](https://github.com/Expensify/App/assets/6833644/67336fd0-4bd6-454b-84a4-be77d245bcb5)

</details>

arosiclair committed:
cache notifications so they can be cleared later

arosiclair committed:
replace unnecessary object params with regular params

arosiclair committed:
remove unused delay functionality

arosiclair committed:
replace object params with regular params

arosiclair committed:
remove unused promise

arosiclair committed:
add reportID to notification data

arosiclair committed:
clear local notifications when opening a report

arosiclair committed:
fix data for modified expense notification

arosiclair committed:
use onclose to clear the cache

arosiclair committed:
use empty object by default for safety

arosiclair committed:
improve docs for LocalNotification lib

arosiclair committed:
clear local notifications when the window is focused

arosiclair commented:
This is working decently well but I'm running into issues with the focus event not being unsubscribed because we're rendering multiple ReportScreens when switching chats. I'm asking about this [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1701121507695419).

arosiclair committed:
prevent background ReportScreens from clearing notifications

arosiclair committed:
undo name change to avoid new JS lib error

arosiclair committed:
prettier

arosiclair committed:
clarifying comments

arosiclair commented:
> I'm running into issues with the focus event not being unsubscribed because we're rendering multiple ReportScreens when switching chats. I'm asking about this [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1701121507695419).

Was able to resolve this by using `isTopMostReportId`. Ready for review!

Julesssss commented on src/hooks/useWindowFocusEvent/useWindowFocusEvent.ts:
```diff
@@ -0,0 +1,19 @@
+ */
+export default function useWindowFocusEvent(callback: UseWindowFocusEventCallback) {
+    useEffect(() => {
+        window.addEventListener('focus', callback);
```
Isn't this getting called twice?

Julesssss replied:
Oh thanks I completely missed the add/remove 😵‍💫

Julesssss commented on src/libs/Notification/LocalNotification/BrowserNotifications.js:
```diff
@@ -1,11 +1,12 @@
 import * as AppUpdate from '@userActions/AppUpdate';
 import focusApp from './focusApp';
 
-const DEFAULT_DELAY = 4000;
+const notificationCache = [];
```
What was the original reason for this delay?

Julesssss replied:
Cool

arosiclair committed:
rename prevNotificationID to notificationID

arosiclair committed:
Merge branch 'arosiclair-clear-read-notifications' into arosiclair-clear-read-notifications-native

arosiclair committed:
Merge branch 'main' of github.com:Expensify/App into arosiclair-clear-read-notifications-native

arosiclair committed:
add android native module for clearing notifications

Julesssss reviewed (dismissed):
LGTM, awaiting C+ review

getusha commented on src/hooks/useWindowFocusEvent/index.website.ts:
```diff
@@ -0,0 +1,3 @@
+import useWindowFocusEvent from './useWindowFocusEvent';
+
+export default useWindowFocusEvent;
```
We can simply move the code inside `useWindowFocusEvent` to `index.ts` and then export the empty function for native from `index.native.ts`, it will make it a bit easier to navigate.

arosiclair replied:
Yes I've seen that in a few places and IMO it's bad practice 😄, but there's nothing in our style guide restricting that so I'll just update this to match 👍

getusha commented on src/hooks/useWindowFocusEvent/index.ts:
```diff
@@ -0,0 +1,9 @@
+ *
+ * @param callback the function to run when focus event is fired for the window. This should be memoized with `useCallback`.
+ */
+export default function useWindowFocusEvent(callback: UseWindowFocusEventCallback) {}
```
i think we can just export `() => {}` here.

arosiclair committed:
Merge branch 'main' of github.com:Expensify/App into arosiclair-clear-read-notifications-native

arosiclair committed:
add clearReportNotifications lib and use airship APIs on native

arosiclair committed:
remove unnecessary native lib

arosiclair commented:
Alright I was going to implement the native side in a separate PR since I thought that would require lots of native code, but I just realized Airship has JS APIs for this so no native code is needed! 

I'm going to finish implementing this for all platforms using just this PR so please HOLD on reviewing @Julesssss @getusha

arosiclair committed:
Revert "rename prevNotificationID to notificationID"

This reverts commit 0f5bd2eb75b17716c4ceb7faeaa99b69c953d50b.

arosiclair committed:
Merge branch 'main' of github.com:Expensify/App into arosiclair-clear-read-notifications

arosiclair committed:
param fixes

arosiclair committed:
convert to typescript

arosiclair committed:
use common types module and default to web implementation

arosiclair committed:
correctly parse the notification and report IDs

arosiclair committed:
delete accidental jsbundle

arosiclair committed:
Merge branch 'main' of github.com:Expensify/App into arosiclair-clear-read-notifications

arosiclair committed:
use new NotificationData type

arosiclair committed:
mock getActiveNotifications to fix tests

arosiclair committed:
remove useless jsdoc

arosiclair committed:
prettier

arosiclair committed:
simplify clearNotifications

arosiclair committed:
use AppState on native

arosiclair committed:
rename to useAppFocusEvent

arosiclair committed:
use airship v15.3.1 patch

fixes Airship.push.getActiveNotifications on Android

arosiclair committed:
Merge branch 'main' of github.com:Expensify/App into arosiclair-clear-read-notifications

arosiclair committed:
use release airship 15.3.1 version

arosiclair committed:
actually use the release version of 15.3.1

arosiclair committed:
update podfile and gemfile

arosiclair commented:
Alright after lots of updates this is ready for review!

- Merged the typescript migration
- Added native implementation for `clearReportNotifications`
- Added native implementation for `useWindowFocusEffect` and renamed it to a more general `useAppFocusEvent`

 @Julesssss I think you'll just need to do a quick test for iOS.

Julesssss commented on src/libs/Notification/clearReportNotifications/index.native.ts:
```diff
@@ -0,0 +1,32 @@
+import ClearReportNotifications from './types';
+
+function parseNotificationAndReportIDs(pushPayload: PushPayload) {
+    let payload = pushPayload.extras.payload;
+    if (typeof payload === 'string') {
+        payload = JSON.parse(payload);
+    }
```
I'm just thinking, but if we end up zipping the payload we'll have to remember to update this too, or will we pass in the already uncompressed payload as a param?

arosiclair replied:
Yeah we have this `JSON.parse` in a couple places so we'd have to update all of those. Not sure what Airship's SDK will do but hopefully it passes the payload uncompressed so we don't have to do anything special

Julesssss commented on src/libs/Notification/clearReportNotifications/index.native.ts:
```diff
@@ -0,0 +1,32 @@
+            .map((notification) => notification.notificationID);
+
+        Log.info(`[PushNotification] found ${reportNotificationIDs.length} notifications to clear`, false, {reportID});
+        reportNotificationIDs.forEach((notificationID) => notificationID && Airship.push.clearNotification(notificationID));
+    });
```
Can these Airship lib functions ever throw exceptions?

Julesssss commented:
> @Julesssss I think you'll just need to do a quick test for iOS.

Would you mind if i did this next Monday? I have a super busy day today trying to [clear the final tasks and release OldDot](https://expensify.slack.com/archives/C07J32337/p1701966422248399).

arosiclair committed:
catch and log unexpected errors

arosiclair committed:
Merge branch 'main' of github.com:Expensify/App into arosiclair-clear-read-notifications

arosiclair committed:
Merge branch 'main' of github.com:Expensify/App into arosiclair-clear-read-notifications

arosiclair commented:
@getusha can you review and finish the rest of the checklist?

getusha commented:
@arosiclair Reassure is failing could you take a look?

arosiclair commented:
> @arosiclair Reassure is failing could you take a look?

Seems to be a flakey test. It's passing now.

s77rt commented:
Taking this as C+

arosiclair committed:
Merge branch 'main' of github.com:Expensify/App into arosiclair-clear-read-notifications

arosiclair commented:
Fixed more merge conflicts. @s77rt added you as a reviewer. Please prioritize this when you get the chance.

s77rt commented on src/libs/Notification/LocalNotification/BrowserNotifications.ts:
```diff
@@ -34,42 +37,37 @@ function canUseBrowserNotifications(): Promise<boolean> {
-    return new Promise((resolve) => {
-        if (!title || !body) {
-            throw new Error('BrowserNotification must include title and body parameter.');
+function push(title: string, body?: string, icon: string | ImageSourcePropType = '', data: LocalNotificationData = {}, onClick: LocalNotificationClickHandler = () => {}) {
```
```suggestion
function push(title: string, body: string, icon: string | ImageSourcePropType = '', data: LocalNotificationData = {}, onClick: LocalNotificationClickHandler = () => {}) {
```

body is required. You can also remove the check inside the function.

arosiclair replied:
Leaving `body` optional here since it's not a strict requirement to show a notification. Removed the unnecessary validation though

s77rt commented on src/libs/Notification/clearReportNotifications/index.native.ts:
```diff
@@ -0,0 +1,38 @@
+    const data = payload as NotificationData;
+    return {
+        notificationID: pushPayload.notificationId,
+        reportID: String(data.reportID),
```
Why we need to cast to string here? Isn't the type on `ReportCommentNotificationData` wrong?

arosiclair replied:
The report ID i'm testing with gets parsed as a `number` so it might be wrong. It's also an old report though. I'll take a closer look at this

arosiclair replied:
Just checked and we are definitely sending `reportID` as `number` from the backend so the type is correct. We just need to cast as string here to make the comparison work

s77rt commented:
Sometimes the focus callback is executed twice (no clear reproduction steps yet)


https://github.com/Expensify/App/assets/16493223/eba3fcc3-08ed-4509-a9a9-334ab6e5ed0c

arosiclair commented on src/hooks/useAppFocusEvent/index.ts:
When I tested it, AppState's change event didn't fire when focusing another window and then re-focusing the app.

s77rt replied:
Great catch!

s77rt commented:
If you switch to the chat (where you'd get messages) and switch to another tab, the notifications won't come in


https://github.com/Expensify/App/assets/16493223/8e5168a4-6bd3-4fa7-b835-d1fbe1b8700d

Julesssss commented:
> Isn't it annoying that you keep getting notifications even when within the app?

I don't think so. You could be in a different chat and then exit the app. In that case I would never see your message.

arosiclair committed:
remove inconsistent param validation

arosiclair committed:
reduce number of array iterations

arosiclair committed:
move native implementation to PushNotification lib

s77rt commented on src/libs/Notification/LocalNotification/types.ts:
```diff
@@ -1,26 +1,16 @@
     showUpdateAvailableNotification: () => void;
-    showModifiedExpenseNotification: (reportCommentParams: ReportCommentParams) => void;
+    showModifiedExpenseNotification: (report: Report, reportAction: ReportAction, onClick: LocalNotificationClickHandler) => void;
+    clearReportNotifications: (reportID: string) => void;
```
```suggestion
    clearReportNotifications: ClearReportNotifications;
```

NAB. Use the exported type from ClearReportNotifications

arosiclair commented:
> If you switch to the chat (where you'd get messages) and switch to another tab, the notifications won't come in

I repro'd this. Looks like there's an issue with `lastReadTime` being updated for the report:

```
[info] [LocalNotification] No notification because the comment was already read - {"created":"2023-12-14 18:17:42.487","lastReadTime":"2023-12-14 18:17:42.818"}
```

It's being updated [here](https://github.com/Expensify/App/blob/arosiclair-clear-read-notifications/src/pages/home/report/ReportActionsList.js#L194) even though the window isn't actually visible

arosiclair committed:
ensure the app is visible before marking the report read

arosiclair committed:
use unified type for ClearReportNotifications

arosiclair committed:
Merge branch 'main' of github.com:Expensify/App into arosiclair-clear-read-notifications

arosiclair committed:
undo accidental change

arosiclair commented:
Alright PR is updated and ready for another review! @s77rt @Julesssss

s77rt commented on src/pages/home/report/ReportActionsList.js:
```diff
@@ -189,12 +190,10 @@ function ReportActionsList({
-                Report.readNewestAction(report.reportID);
-            } else {
-                readActionSkipped.current = true;
-            }
+        if (Visibility.isVisible() && ReportUtils.isUnread(report) && scrollingVerticalOffset.current < MSG_VISIBLE_THRESHOLD) {
+            Report.readNewestAction(report.reportID);
+        } else {
+            readActionSkipped.current = true;
```
Can you please re-add the inner condition. `readActionSkipped` should be set to true only if the report is unread.

s77rt commented:
## Reviewer Checklist

- [X] I have verified the author checklist is complete (all boxes are checked off).
- [X] I verified the correct issue is linked in the `### Fixed Issues` section above
- [X] I verified testing steps are clear and they cover the changes made in this PR
    - [X] I verified the steps for local testing are in the `Tests` section
    - [X] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [X] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [X] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I verified tests pass on **all platforms** & I tested again on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/8307fe49-31d5-4f9f-a536-59d858755e7d


</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->
Not applicable

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->
Can't test notifications on iOS



</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->
Not applicable

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/333c97d5-4e01-4a9f-bb32-dbfff31e23ae

https://github.com/Expensify/App/assets/16493223/66692936-7885-47c2-b291-9e8b00193df4


</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/1a52167b-a933-4038-b6b7-9fc4ef5c622b


</details>

arosiclair committed:
set readActionSkipped only when report is unread

marcochavezf reviewed (approved):
The code LGTM

arosiclair commented:
Alright I tried getting an iOS tester a couple times [here](https://expensify.slack.com/archives/C03TQ48KC/p1702316916424349) with no luck. Since this is on a time crunch I just re-tested myself here:

https://github.com/Expensify/App/assets/6833644/4792a554-6956-458a-9529-b34276cf4712

Looking good so going to merge

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

arosiclair commented:
Just realized I introduced another require cycle:

```
 WARN  Require cycle: src/libs/actions/Report.ts -> src/libs/Notification/clearReportNotifications/index.native.ts -> src/libs/Notification/PushNotification/index.native.ts -> src/libs/Notification/PushNotification/ForegroundNotifications/index.ios.ts -> src/libs/Notification/PushNotification/shouldShowPushNotification.ts -> src/libs/actions/Report.ts
```

Opened a new issue to address that [here](https://github.com/Expensify/App/issues/33204)

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7256301720) to staging by https://github.com/arosiclair in version: 1.4.14-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7256301720) to staging by https://github.com/arosiclair in version: 1.4.14-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7284210064) to production by https://github.com/jasperhuangg in version: 1.4.14-6 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

QichenZhu commented on src/pages/home/report/ReportActionsList.js:
```diff
@@ -190,7 +191,7 @@ function ReportActionsList({
 
         if (ReportUtils.isUnread(report)) {
-            if (scrollingVerticalOffset.current < MSG_VISIBLE_THRESHOLD) {
+            if (Visibility.isVisible() && scrollingVerticalOffset.current < MSG_VISIBLE_THRESHOLD) {
                 Report.readNewestAction(report.reportID);
```
On desktop web, notifications stopped working after the user switched to another app, since `document.visibilityState` is [`'visible'`](https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilityState#visible) when the browser loses focus but is neither minimized nor on another tab.

Fixed via #93387 by also checking [`document.hasFocus()`](https://developer.mozilla.org/en-US/docs/Web/API/Document/hasFocus).
